### PR TITLE
Fix stack-dependent related_integrations.version export

### LINKED ISSUE #5601: [Bug] Improper usage of the rule version
## Summary

The primary and backport **security_detection_engine** packages reuse the same prebuilt rule version. It leads to the bug when user sees a warning regarding related integrations version mismatch but there are no rule upgrades available to be installed.

## Details

**Let's consider the issue on the particular example**

- We have `Lateral Movement via Startup Folder` prebuilt rule in **security_detection_engine** package for `8.19.x` and `9.3.x` stack versions
- User installs `Lateral Movement via Startup Folder` version `312` from the **security_detection_engine** package for `8.19.x` stack version ([rule's JSON](https://github.com/elastic/integrations/blob/backport-security_detection_engine-8.19/packages/security_detection_engine/kibana/security_rule/25224a80-5a4a-4b8a-991e-6ab390465c4f_312.json#L25-L42)). It has `endpoint` integration dependency specified as `^8.2.0`.
- User upgrades Kibana to the recent stack version, let's say it's `9.3.x`.
- User doesn't see any rule upgrades as `Lateral Movement via Startup Folder` in the **security_detection_engine** package for `9.3.x` stack version has the same version `312`. Effectively the installed prebuilt rule is from `8.19.x`.
-  `Lateral Movement via Startup Folder` in the **security_detection_engine** package for `9.3.x` stack version has `related_integration` field specifying `endpoint` integration dependency specified as `^9.0.0` (`endpoint` integration dependency specified as `^8.2.0`).

In result the user sees the prebuilt rule from the older stack version but there is no way to upgrade it to update the `related_integrations` field. The only way to fix it from the user side is to remove the prebuilt rule and install it again.

### PR DESCRIPTION
**Pull request by Mikaayenson**
*Issue link(s)*: Resolves #5601

## Summary - What I changed

This PR fixes several bugs in how `related_integrations.version` is computed at export time. All of them affected the value written into prebuilt rules; none change ES|QL validation (that still uses `find_latest_compatible_version` / patch inference from #6251). Addressing the initial bug is a smaller PR, but through testing, we identified several other issues predominantly with ESQL (_given we do not have an ESQL parser_), that expands the scope of the changes.

### Bugs fixed

1. **Stack-dependent export (#5601).** `find_least_compatible_version` picked a single `^X.Y.Z` from the build-time stack version. The same rule id could ship `^8.2.0` on an 8.19 backport and `^9.0.0` on 9.x while the prebuilt rule `version` integer stayed the same, which triggered false integration version-mismatch warnings after upgrade.

2. **Schema-blind OR ranges.** Rebased onto #6251 so export respects `integration-schemas.json.gz`: data streams introduced mid-package (for example `aadgraphactivitylogs` in azure 1.37.0) get a schema floor in the OR range, not just Kibana manifest compatibility.

3. **Phantom anchors from unshipped stack lines.** Walking every Kibana major in the manifest could emit legacy anchors such as `^0.0.2` from azure packages that still declare `^7.9.0`. Export now intersects with stack majors from the stack-schema-map backport lines we actually ship rules to.

4. **Incomplete stack-major coverage.** Bounded ranges like `>=8.12.0 <9.1.0` omitted major 9; unbounded `>=` ranges stopped at the lower bound only. Fixed with per-clause major overlap and a cap for open-ended ranges.

5. **Duplicate ES|QL `related_integrations` rows.** ES|QL rules tagged `[metadata] integration = ["azure"]` with a query filter on `azure.signinlogs` (or similar) got both a spurious metadata-only package row and the data-stream row. Kuery/EQL were unaffected. Fixed with a prefix check so the metadata row is skipped when the query already names a dataset for that package.

### Implementation

- Replaces `find_least_compatible_version` with `find_compatible_version_range`, which returns a stack-invariant OR'd caret range (for example `^8.2.0 || ^9.0.0 || ^10.0.0` for endpoint) plus a forward sentinel for the next integration major.
- `policy_templates` are unioned across manifest-backed anchors, not parsed from the OR string.
- `UNKNOWN_PACKAGE_INTEGRATION` constant; tuple anchors on `CompatibleVersionRange`.

**Performance:** the new path resolves more stack lines per package. Informal local timing was roughly 2 to 9x median vs the old function on `endpoint` / `aws` / `windows` manifests, still sub-10ms per call and once per rule at package build. Not expected to move `build-release` wall time in a meaningful way.

**Follow-up:** `version.lock` bumps for affected rules are planned separately before the next package release (per review).

## How To Test

```bash
make deps
make pytest
python -m detection_rules view-rule rules_building_block/defense_evasion_download_susp_extension.toml
```

Spot check export: rule `8d366588-cbd6-43ba-95b4-0971c3f906e5` should show `^8.2.0 || ^9.0.0 || ^10.0.0` for endpoint on every release line.

```bash
  "related_integrations": [
    {
      "package": "endpoint",
      "version": "^8.2.0 || ^9.0.0 || ^10.0.0"
    }
  ],
```

Export toml and import into kibana to ensure it applies successfully.

```bash
uv run python -m detection_rules export-rules-from-repo -f rules_building_block/defense_evasion_download_susp_extension.toml
```

[20260528T154018L.ndjson.txt](https://github.com/user-attachments/files/28367569/20260528T154018L.ndjson.txt)

Screenshot 2026-05-28 at 3 42 57 PM

Screenshot 2026-05-28 at 3 46 47 PM

Backport validation: cherry-pick this PR onto each release branch (or apply the changed files from the branch), then run the same pytest commands. No `version.lock` changes required for that pass.

<details><summary>Backport `8.19` (origin/8.19 + PR code)</summary>

```
$ pytest -q
FAILED tests/test_all_rules.py::TestRuleMetadata::test_rule_change_has_updated_date
1 failed, 221 passed, 19 skipped, 2 subtests passed in 69.53s (0:01:09)

# same failure on vanilla origin/8.19 (diffs rules/ vs origin/main, not introduced here)

$ pytest -q -k "not test_rule_change_has_updated_date"
221 passed, 19 skipped, 1 deselected, 2 subtests passed in 65.20s (0:01:05)

$ pytest tests/test_integrations.py::TestFindCompatibleVersionRange -q
.......                                                                  [100%]
7 passed in 0.51s

$ spot-check 8d366588-cbd6-43ba-95b4-0971c3f906e5
[{'package': 'endpoint', 'version': '^8.2.0 || ^9.0.0 || ^10.0.0'}]
```

</details>

<details><summary>Backport `9.2` (origin/9.2 + PR code)</summary>

```
$ pytest -q
222 passed, 19 skipped, 2 subtests passed in 82.76s (0:01:22)

$ pytest tests/test_integrations.py::TestFindCompatibleVersionRange -q
.......                                                                  [100%]
7 passed in 0.54s

$ spot-check 8d366588-cbd6-43ba-95b4-0971c3f906e5
[{'package': 'endpoint', 'version': '^8.2.0 || ^9.0.0 || ^10.0.0'}]
```

</details>

<details><summary>Backport `9.3` (origin/9.3 + PR code)</summary>

```
$ pytest -q
222 passed, 19 skipped, 2 subtests passed in 96.19s (0:01:36)

$ pytest tests/test_integrations.py::TestFindCompatibleVersionRange -q
.......                                                                  [100%]
7 passed in 0.52s

$ spot-check 8d366588-cbd6-43ba-95b4-0971c3f906e5
[{'package': 'endpoint', 'version': '^8.2.0 || ^9.0.0 || ^10.0.0'}]
```

</details>

<details><summary>Backport `9.4` (origin/9.4 + PR code)</summary>

```
$ pytest -q
222 passed, 19 skipped, 2 subtests passed in 106.83s (0:01:46)

$ pytest tests/test_integrations.py::TestFindCompatibleVersionRange -q
.......                                                                  [100%]
7 passed in 0.50s

$ spot-check 8d366588-cbd6-43ba-95b4-0971c3f906e5
[{'package': 'endpoint', 'version': '^8.2.0 || ^9.0.0 || ^10.0.0'}]
```

</details>

## Checklist

- [x] Added a label for the type of pr: `bug`
- [ ] Added the `meta:rapid-merge` label if planning to merge within 24 hours
- [x] Secret and sensitive material has been managed correctly
- [x] Automated testing was updated or added to match the most common scenarios
- [ ] Documentation and comments were added for features that require explanation

### CODE REVIEW

File: detection_rules/integrations.py [hunk 1]
```diff
@@ -244,36 +245,165 @@ def _satisfies_kibana_range(stack: Version, version_requirement: str) -> bool:
...
+
+def _stack_majors_supported_by_package(integration_manifests: dict[str, Any]) -> set[int]:
+    """Collect Kibana stack majors that any manifest in the package can serve."""
+    stack_majors: set[int] = set()
+    for manifest in integration_manifests.values():
+        version_requirement = manifest["conditions"]["kibana"]["version"]
+        for lo, hi in _parse_kibana_range(version_requirement):
+            majors_to_check: list[int]
+            if hi is None:
+                majors_to_check = [lo.major]
```
Comment (eric-forte-elastic) [inline detection_rules/integrations.py]: For this if check:
>fix(integrations): tighten stack-major overlap and anchor resolution
Walk every stack major whose band intersects a bounded Kibana clause
(e.g. >=8.12.0 <9.1.0 includes major 9) and pick the earliest compatible
stack point within a major for the legacy least-compatible walk.

I don't think any of our manifests include a situation where `hi is None` unless one were to have a manifest with an unbounded range (e.g. ">=8.12.0"). Fine to have this check, but given this I am not sure that this functions as intended. 

For the case with an unbounded range `majors_to_check` only will have `[8]` and not any other stack majors. Not sure if this accomplishes the goal. 

Image

 Whereas for the case in the note, it correctly has multiple stack versions to check. 

Image

**Comment by eric-forte-elastic** [comment]
Validated expected results from testing/debugging. :green_circle: base flows function as designed.

<details><summary>Details</summary>
<p>

image

image

image


image



</p>
</details>

File: detection_rules/integrations.py [hunk 2]
```diff
@@ -244,36 +245,165 @@ def _satisfies_kibana_range(stack: Version, version_requirement: str) -> bool:
...
+    range: str
+    anchors: list[str]
+    forward_anchor: str
+
+
+def find_compatible_version_range(
+    package: str,
+    packages_manifest: dict[str, Any],
+) -> CompatibleVersionRange:
+    """Return a stack-invariant OR'd caret range for related_integrations.version.
```
Comment (eric-forte-elastic) [inline detection_rules/integrations.py]: Nit. Should this doc string be updated to fit the existing format? Short one sentence doc string with more detail in separate comment?

File: tests/test_integrations.py [hunk 3]
```diff
@@ -215,45 +215,88 @@ def test_unknown_package_raises(self):
...
+                "2.0.0": _manifest("^2.0.0"),
             }
         }
-        with self.assertRaises(ValueError):
-            find_least_compatible_version("pkg", "pkg", "9.1.0", manifests)
+        first = find_compatible_version_range("pkg", manifests)
+        second = find_compatible_version_range("pkg", manifests)
+        self.assertEqual(first, second)
+
+    def test_single_major_appends_forward_anchor(self):
```
Comment (eric-forte-elastic) [inline tests/test_integrations.py:254]: Nit. Akin to https://github.com/elastic/detection-rules/pull/6208/changes#r3325352838, if the goal is to support unbounded upper ranges like `">=8.12.0"` should we include a test for that? The `hi is none` branch?

**Review by eric-forte-elastic — APPROVED** [review]
[One main comment](https://github.com/elastic/detection-rules/pull/6208/changes#r3325352838) and a few nits. Otherwise LGTM 👍 

Generally this PR is low risk, as the updated version computation logic in effect is not used in other workflows (like schema validation, etc.). Additionally, Kibana validates this to be a NonEmpty string just like we do in our dataclasses. ([ref](https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.gen.ts#L559)), so we have considerable freedom in what we populate it with.

File: detection_rules/integrations.py (hunk 2 above)
Comment (shashank-elastic) [reply detection_rules/integrations.py]: +1 for this cleanup

**Review by shashank-elastic — APPROVED** [review]
This does not affect the release as such , but lock versions are impacted. Since we will handle that separately before the release ( either bump 1 across all or selectively bump 1 on affected rules), this PR should be good to merge. 

Also checked CLI paths affected. It is working as expected. 

```console
python -m detection_rules dev integrations show-latest-compatible -p elastic_security -s 9.4.0          
Loaded config file: /Users/shashankks/elastic_workspace/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Compatible integration version=('0.4.0', [''])

detection-rules on  5601-bug-improper-prebuilt-rule-version-usage-in-security_detection_engine-package-versions [$?] is 📦 v1.6.44 via 🐍 v3.12.8 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ python -m detection_rules dev integrations show-latest-compatible -p pad -s 9.4.0             
Loaded config file: /Users/shashankks/elastic_workspace/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Compatible integration version=('2.0.0', [''])

detection-rules on  5601-bug-improper-prebuilt-rule-version-usage-in-security_detection_engine-package-versions [$?] is 📦 v1.6.44 via 🐍 v3.12.8 (.venv) on ☁️  shashank.suryanarayana@elastic.co 
❯ 
```

**Comment by eric-forte-elastic** [comment]
New deduplication logic make sense and functions as expected: 
Old: 
image
New:
image
